Track layer slider max value

horsto opened the pull request:
# References and relevant issues

Closes https://github.com/napari/napari/issues/7728

# Description
This PR takes care of ~~two~~ one thing: 
- In case users set a new `.tail_length` (or `.head_length`) on an existing tracks layer that is higher than the standard value of 300, display the track lengths correctly and adjust the new max of the tail or head length sliders to that new value

The below part was part of the original PR, but I scratched it since developers decided it would be best to separate it into a new project / PR in the future. 
~~I noticed that the shader tapers off alpha value of the earliest visible track points (points at the very beginning of tail or end of head) to 0 (zero). This counteracts the user's input because at a tail length of 500 for example, the earliest 100-200 points are barely visible. I clamped the alpha value to 0.5 instead of 0, which I feel is more reasonable because it leads to a "slide fade" (does not get rid of the effect completely), but maintains some visibility. One could think about enabling or disabling this feature with a checkbox. 
Below is a direct comparison. Note the change in visibility for blue (early) track points.~~

```
### Alpha zero (original)
![alpha 0](https://github.com/user-attachments/assets/c5171aba-753d-4878-97a2-c2c493a1344b)

### Alpha clamped to 0.5 (new)
![alpha 5 clamped](https://github.com/user-attachments/assets/784a120a-3260-4a65-a505-59ab864b0590)

```

horsto committed:
Adjust tail and head_length_slider when value>max

This fixes the issue that occurs when setting a new tail length that is longer than the current maximum allowed on either the head or tail length slider. It sets the slider to the new max.

horsto committed:
Update visibility of early time values in tracks

Tracks fade out of view completely at early timepoints (alpha tapers off to 0). Clamping alpha to .5 to have "mild" fading only.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

TimMonko commented on napari/_qt/layer_controls/qt_tracks_controls.py:
Looks and works great!

TimMonko commented on napari/_vispy/filters/tracks.py:
Personally, I would take this alpha change out and make a separate PR, but definitely take the advice of the core devs. I partly suggest that because I'm unsure about the change, but also because then each PR will serve a single 'task'. 

For me, having the alpha fade completely out is very nice when there are lots of competing tracks in the canvas. Try out the [tracks_3d example](https://github.com/horsto/napari/blob/track_layer_slider+opacity/examples/tracks_3d.py). With this alpha it's a bit too cluttered for me, even with the normal tail length. 

That said I'd be very happy if the alpha was settable with something like `fade_alpha`-- but that's much more work to hook up😬 

Perhaps try a different LUT than turbo. Something like PiYG will never 'fade to black', same with the inverted cmaps like 'I Purple', so may look nicer with long tails for you. What do you think with those? 

I totally support and understand why you've changed the alpha regardless! 

As an aside ... I'm surprised turbo is the default?

horsto replied:
Good points with various colormaps giving various amounts of "visible" fading (I did set it to `turbo` myself). I do think that introducing a "fade tracks" checkbox or so would be nice, simply because if, as a user, I demand 1000 frames track tail, I want to see all 1000 frames. Totally see the point of fading and I like the effect, but I think there are pros and cons for both.

horsto replied:
> I would take this alpha change out and make a separate PR 

Yes, happy to do that (sorry I didn't mention it in my previous reply). Waiting for advice from others too.

DragaDoncila replied:
Thanks for the pr @horsto! I agree with Tim that we should split out the alpha change from this PR, and I'm leaning towards including a `fade_alpha` or `fade_opacity` checkbox or even slider that allows the user to set the minimum alpha value.

horsto replied:
Great, ok, I deleted the alpha changing commit from this PR.

horsto replied:
And updated the description accordingly.

TimMonko reviewed (commented):
@horsto grats on the PR! You've got many great ideas so I hope you keep contributing! 

With this you can update the description to 'Closes #7728' and it will automatically close that issue if this is merged. 

🚀

horsto committed:
Revert "Update visibility of early time values in tracks"

This reverts commit 82cacf0f08a0ba9dc31c1d0b7e9e6e57fbcdd4ba.

horsto committed:
Merge branch 'track_layer_slider+opacity' of https://github.com/horsto/napari into track_layer_slider+opacity

DragaDoncila reviewed (approved):
Awesome, thanks for that change @horsto :tada: It would be great if you could open an issue to track the request for setting a min alpha for the opacity fading. I'm approving and  I've also changed the milestone to 0.6.0 since this is a small and non-controversial change. Congrats on your first contribution! :fire:

TimMonko reviewed (approved):
:rocket:

horsto committed:
Add tests

Added test_update_max_tail_length() and test_update_max_head_length() according to suggestion from @DragaDoncila . Unified QtTracksControls(layer) variable naming (controls) across testing functions.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
